### RandomForest Feature Importance for Selection

Process steps:
* Build full dataset with decoys and compounds
* Build compound-only dataset
* Preprocess categorical with one-hot-encoding
* Find intersecting columns and filter compound-only to these
* Add non-linear transformations to compound-only dataset
* Create binary target with IC50 <= 10 as 1, else 0
* Using compound-only df, fine-tune random forest with GridSearchCV (CV = count of positive class)
* Extract feature importance and remove any with importance of 0
* Create important features in full dataset

In [2]:
import pandas as pd
import numpy as np
from scipy.stats import skewtest
import os
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import cross_val_predict
from sklearn.model_selection import GridSearchCV
from sklearn.preprocessing import Imputer
from sklearn.metrics import confusion_matrix, log_loss
from custom_functions import *

In [3]:
avail_transformations = ["log", "log2", "log10", "cubert", 
                         "sqrt", "exp", "exp2", "cube", "sq"]

* Build full-dataset
* Build compound-dataset
* Preprocess data adding one-hot-encoded features for both
* Find intersecting columns
* Add non-linear transformations and drop na's

In [4]:
# Load in full dataset
full_x, full_y = load_full_dataset()
# Preprocess variables
full_x = preprocess_variables(full_x)
# Extract list of available columns
full_columns = full_x.columns
print("Loading in compound dataset....")
# Read in compound dataset
compound_x, compound_y = load_compound_dataset()
# Preprocess
compound_x = preprocess_variables(compound_x)
# Find intersecting features
avail_columns = compound_x.columns.intersection(full_columns)
# Select features on subset
x_data = compound_x.loc[:, avail_columns]
y_data = compound_y.copy()
print("Adding non-linear features to compound dataset....")
# Add all transformations on compound data
for feature in x_data.columns[x_data.dtypes == 'float64']:
    x_data = add_transformations(x_data, feature)
# Drop any new columns with NaN due to improper transformation
x_data.replace([np.inf, -np.inf], np.nan, inplace=True)
x_data.dropna(axis=1, inplace=True)
assert not sum(x_data.isna().sum()), "Unexpected nulls found"

Adding Akt1_decoys_padel.csv....
Adding AmpC_decoys_padel.csv....
Adding cp3a4_decoys_padel.csv....
Adding cxcr4_decoys_padel.csv....
Adding gcr_decoys_padel.csv....
Adding HIVpr_decoys_padel.csv....
Adding HIVrt_decoys_padel.csv....
Adding Kif11_decoys_padel.csv....
Loading in compound dataset....
Adding non-linear features to compound dataset....


In [5]:
# Create binary variable
y_class = np.squeeze([int(y_val <= 10) for y_val in y_data])

In [6]:
# How well does random forest predict potency?
print("Tuning random forest on compound dataset....")
model = RandomForestClassifier(random_state=0, 
                               class_weight={0: 1-np.mean(y_class), 
                                             1: np.mean(y_class)})
params = {"n_estimators": [30, 40, 50, 60],
          "max_depth": [3, 4, 5],
          "min_samples_split": [2, 3],
         "criterion": ["entropy"],
         "max_features": ["auto", "sqrt", "log2"]}
grid = GridSearchCV(estimator=model, param_grid=params, cv=5, n_jobs=3)

grid.fit(x_data, y_class)
print(grid.best_params_)
best_model = grid.best_estimator_

Tuning random forest on compound dataset....
{'max_depth': 4, 'max_features': 'auto', 'criterion': 'entropy', 'n_estimators': 40, 'min_samples_split': 2}


In [7]:
# Analyze CV prediction performance
predict = cross_val_predict(
    best_model, x_data, y_class, cv=sum(y_class), method="predict")

predict_proba = cross_val_predict(
    best_model, x_data, y_class, cv=sum(y_class), method="predict_proba")

print(confusion_matrix(y_class, predict, labels=[1, 0]))
print(np.array([["TP", "FN"], ["FP", "TN"]]))


[[ 3  8]
 [ 4 32]]
[['TP' 'FN']
 ['FP' 'TN']]


In [8]:
# Analyze CV Performance
pd.DataFrame(
    {"IC50": y_data, 
     "y_class": y_class, 
     "Prediction": predict, 
     "Proba": predict_proba[:,1]})[
    ["IC50", "y_class", "Prediction", "Proba"]]

IC50  y_class  Prediction     Proba
0     0.036        1           0  0.325236
1    10.000        1           0  0.200000
2    50.000        0           0  0.425000
3    50.000        0           0  0.475000
4    50.000        0           0  0.350236
5     8.000        1           1  0.555851
6    50.000        0           0  0.330851
7    50.000        0           0  0.102651
8    35.000        0           0  0.151211
9    50.000        0           0  0.251211
10   45.000        0           0  0.102651
11   45.000        0           0  0.100280
12   40.000        0           0  0.425280
13   50.000        0           0  0.405851
14   50.000        0           0  0.175280
15   50.000        0           0  0.050920
16   50.000        0           0  0.100000
17   25.000        0           0  0.155851
18   50.000        0           1  0.625000
19   50.000        0           0  0.300000
20   15.000        0           1  0.679231
21    1.700        1           0  0.250000
22   10.000        1           1  0.604231
23   10.000        1           1  0.825000
24   10.000        1           0  0.375000
25    2.000        1           0  0.100000
26   10.000        1           0  0.450261
28   10.000        1           0  0.000261
48   50.000        0           1  0.925000
52    9.400        1           0  0.425000
53   50.000        0           1  0.800000
54  100.000        0           0  0.125000
55  100.000        0           0  0.430851
56   17.600        0           0  0.150000
57  100.000        0           0  0.100000
58  100.000        0           0  0.275000
59  100.000        0           0  0.205851
60  100.000        0           0  0.075000
61  100.000        0           0  0.150261
62   25.000        0           0  0.178574
63   50.000        0           0  0.063057
64   50.000        0           0  0.025261
65   50.000        0           0  0.075261
66   50.000        0           0  0.050261
67   50.000        0           0  0.226211
68   50.000        0           0  0.201211
69   18.500        0           0  0.176211

In [9]:
# Analyze train prediction performance
best_model.fit(x_data, y_class)
predict = best_model.predict(x_data)
print(confusion_matrix(y_class, predict, labels=[1, 0]))
print(np.array([["TP", "FN"], ["FP", "TN"]]))

[[11  0]
 [ 0 36]]
[['TP' 'FN']
 ['FP' 'TN']]


In [10]:
# Which ones did it get without CV?
pd.DataFrame({"IC50": y_data, "y_class": 
              y_class, "Prediction": predict})

IC50  Prediction  y_class
0     0.036           1        1
1    10.000           1        1
2    50.000           0        0
3    50.000           0        0
4    50.000           0        0
5     8.000           1        1
6    50.000           0        0
7    50.000           0        0
8    35.000           0        0
9    50.000           0        0
10   45.000           0        0
11   45.000           0        0
12   40.000           0        0
13   50.000           0        0
14   50.000           0        0
15   50.000           0        0
16   50.000           0        0
17   25.000           0        0
18   50.000           0        0
19   50.000           0        0
20   15.000           0        0
21    1.700           1        1
22   10.000           1        1
23   10.000           1        1
24   10.000           1        1
25    2.000           1        1
26   10.000           1        1
28   10.000           1        1
48   50.000           0        0
52    9.400           1        1
53   50.000           0        0
54  100.000           0        0
55  100.000           0        0
56   17.600           0        0
57  100.000           0        0
58  100.000           0        0
59  100.000           0        0
60  100.000           0        0
61  100.000           0        0
62   25.000           0        0
63   50.000           0        0
64   50.000           0        0
65   50.000           0        0
66   50.000           0        0
67   50.000           0        0
68   50.000           0        0
69   18.500           0        0

Analyze feature importance

In [11]:
best_model.fit(x_data, y_class)
feat_importance = best_model.feature_importances_
best_features = [
    (f, i) for i, f in sorted(zip(feat_importance, x_data.columns), 
                              reverse=True) if i != 0]
feat_df = pd.DataFrame(best_features)
feat_df.columns = ["Feature", "Importance"]
feat_df

Feature  Importance
0             AATSC0i_cube    0.025567
1               MATS7c_exp    0.018404
2                   ATSC5m    0.017657
3              MATS7c_cube    0.016712
4              ATSC0i_log2    0.014479
5                   GATS4i    0.014219
6                AATS0p_sq    0.013451
7              MATS1m_exp2    0.013293
8             AATSC5m_cube    0.013293
9                    Si_sq    0.013252
10                  WTPT-5    0.012698
11              AATSC0i_sq    0.012676
12             MATS1m_cube    0.012090
13             MATS5p_exp2    0.012031
14              AATS6i_log    0.011917
15                  AATS0i    0.011917
16              MATS4s_exp    0.011312
17            AATSC5p_exp2    0.011095
18   McGowan_Volume_cubert    0.011095
19                Mpe_cube    0.010959
20             AATSC7e_exp    0.010625
21            AATSC6e_exp2    0.010521
22            SpMAD_D_cube    0.010467
23               SIC4_cube    0.010282
24                 MLFER_E    0.010270
25             WTPT-5_cube    0.010270
26             ATSC1p_exp2    0.010186
27               JGI5_exp2    0.010119
28             GATS1v_exp2    0.009974
29               AATS7p_sq    0.009453
..                     ...         ...
114            piPC7_log10    0.003819
115            MATS7m_exp2    0.003744
116            ATSC5i_cube    0.003679
117          AATS3m_cubert    0.003671
118          AATSC2i_log10    0.003456
119            AATS4p_exp2    0.003353
120           AATSC6i_cube    0.003247
121            AATS7e_sqrt    0.003221
122            GATS5c_log2    0.003093
123            AATSC3c_exp    0.002974
124             ECCEN_1006    0.002953
125            GATS2p_sqrt    0.002931
126              CIC3_log2    0.002901
127           AATS0m_log10    0.002878
128             GATS1s_log    0.002851
129              CIC2_exp2    0.002665
130           AATSC5p_cube    0.002616
131             ATS5i_sqrt    0.002534
132              AATS0v_sq    0.002529
133          ATSC2i_cubert    0.002454
134             AATS2e_log    0.002347
135              MATS2m_sq    0.002252
136             ATS5p_sqrt    0.002167
137           VR2_Dt_log10    0.002101
138           AATSC3e_cube    0.001981
139           AATSC6p_cube    0.001912
140                 MATS4e    0.001783
141         SpAD_Dt_cubert    0.001764
142               FMF_exp2    0.001403
143          GATS8m_cubert    0.001396

[144 rows x 2 columns]

**Analysis:**

When using LOOCV (leaving out one positive-class example each split), 3 relatively-potent compounds are identified. This shows promise that at least some of the characteristics associated with potency are discovered and leveraged for prediction.

The more potent compounds, however, are not identified through the prediction when using cross-validation. This would indicate that the primary characteristics of potency discovered thus-far are not leveraged in this predictive model.

When analyzing the train error on the compound dataset, the model overfits and accurately predicts the potency class of every compound. The information this can provide us with is what features are used to make this distinction. We should try to leverage the features identified here, and through other feature selection techniques, as most important for further exploration and use as a basis for the data augmentation analysis when analyzing the impact of having the decoys added to the dataset.

We can also try to leverage these same features (144 important features) in different models to see if they are able to find better seperation between potent/non-potent when using LOOCV (one split per positive example).